# Bài 1: Làm quen với PyTorch

## 1. Mục tiêu
- Tạo, kiểm tra và thao tác với Tensor trong PyTorch.
- Thực hiện indexing, slicing, reshape và các phép toán ma trận trên Tensor.
- Sử dụng các hàm thống kê cơ bản trên Tensor.
- Hiểu cơ chế automatic differentiation (autograd) và vai trò của gradient.
- Sử dụng optimizer để giải bài toán tối ưu đơn giản bằng Gradient Descent.

> PyTorch cung cấp hai khả năng nền tảng: tính toán Tensor (tương tự NumPy, có thể tăng tốc bằng GPU) và automatic differentiation để huấn luyện mô hình.


In [2]:
import torch
import numpy as np
print("torch:", torch.__version__)
print("cuda available:", torch.cuda.is_available())


torch: 2.14.0+cpu
cuda available: False


---
## Bài 1. Tạo Tensor cơ bản
Tạo một Tensor số nguyên kích thước 2 x 3 và hiển thị kết quả.

**Yêu cầu:** Cho biết kích thước, số chiều và kiểu dữ liệu của Tensor `a`.


In [3]:
import torch

a = torch.tensor(
    [[1, 2, 3],
     [4, 5, 6]],
    dtype=torch.int32
)
print(a)

# --- Tra loi yeu cau ---
print("Shape (kich thuoc):", a.shape)
print("So chieu (ndim):", a.ndim)
print("Kieu du lieu (dtype):", a.dtype)
print("So phan tu (numel):", a.numel())


tensor([[1, 2, 3],
        [4, 5, 6]], dtype=torch.int32)
Shape (kich thuoc): torch.Size([2, 3])
So chieu (ndim): 2
Kieu du lieu (dtype): torch.int32
So phan tu (numel): 6


**Nhận xét Bài 1:**
- `shape = torch.Size([2, 3])`: 2 hàng, 3 cột.
- `ndim = 2`: Tensor 2 chiều (ma trận).
- `dtype = torch.int32`: số nguyên 32-bit.


---
## Bài 2. Tạo các Tensor thường dùng
Thực hành tạo Tensor ngẫu nhiên, Tensor toàn số 0, toàn số 1 và ma trận đơn vị.

**Yêu cầu:**
- Nhận xét sự khác nhau giữa `torch.rand()` và `torch.randn()`.
- Tạo thêm một Tensor 5 x 5 chứa toàn giá trị 10.


In [4]:
import torch
torch.manual_seed(0)  # de ket qua co the tai lap

a = torch.rand(3, 4)
b = torch.randn(3, 4)
c = torch.zeros(2, 3)
d = torch.ones(2, 3)
e = torch.eye(4)

print("a = torch.rand(3, 4):")
print(a)
print("\nb = torch.randn(3, 4):")
print(b)
print("\nc = torch.zeros(2, 3):")
print(c)
print("\nd = torch.ones(2, 3):")
print(d)
print("\ne = torch.eye(4):")
print(e)

# --- Yeu cau them: Tensor 5x5 toan gia tri 10 ---
f = torch.full((5, 5), 10)
print("\nf = Tensor 5x5 toan gia tri 10:")
print(f)


a = torch.rand(3, 4):
tensor([[0.4963, 0.7682, 0.0885, 0.1320],
        [0.3074, 0.6341, 0.4901, 0.8964],
        [0.4556, 0.6323, 0.3489, 0.4017]])

b = torch.randn(3, 4):
tensor([[ 0.4033,  0.8380, -0.7193, -0.4033],
        [-0.5966,  0.1820, -0.8567,  1.1006],
        [-1.0712,  0.1227, -0.5663,  0.3731]])

c = torch.zeros(2, 3):
tensor([[0., 0., 0.],
        [0., 0., 0.]])

d = torch.ones(2, 3):
tensor([[1., 1., 1.],
        [1., 1., 1.]])

e = torch.eye(4):
tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])

f = Tensor 5x5 toan gia tri 10:
tensor([[10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10]])


**Nhận xét Bài 2:**
- `torch.rand()`: phân phối **đều (uniform)** trong khoảng [0, 1). Giá trị luôn >= 0 và < 1.
- `torch.randn()`: phân phối **chuẩn (normal)** trung bình 0, phương sai 1, N(0,1). Giá trị có thể âm/dương, thường trong [-3, 3].
- `zeros / ones / eye / full`: tạo Tensor hằng, hữu ích để khởi tạo trọng số, bias, ma trận đơn vị.


---
## Bài 3. Kiểm tra thông tin Tensor
Sử dụng `shape`, `size()`, `ndim` và `dtype` để kiểm tra các thuộc tính của Tensor.

**Yêu cầu:** Giải thích ý nghĩa của `torch.Size([2, 3, 4])`.


In [5]:
import torch

a = torch.zeros(2, 3, 4)
print("Tensor:")
print(a)
print("Shape:", a.shape)
print("Size:", a.size())
print("Number of dimensions:", a.ndim)
print("Data type:", a.dtype)
print("So phan tu:", a.numel())


Tensor:
tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])
Shape: torch.Size([2, 3, 4])
Size: torch.Size([2, 3, 4])
Number of dimensions: 3
Data type: torch.float32
So phan tu: 24


**Giải thích Bài 3:**
- `torch.Size([2, 3, 4])` nghĩa là Tensor có **3 chiều** (ndim=3):
  - Chiều 0 có độ dài 2 (2 khối),
  - Chiều 1 có độ dài 3 (mỗi khối 3 hàng),
  - Chiều 2 có độ dài 4 (mỗi hàng 4 cột).
- Tổng số phần tử = 2 x 3 x 4 = 24.
- `shape` và `size()` trả về cùng giá trị; `shape` là thuộc tính, `size()` là phương thức.


---
## Bài 4. Slicing và Indexing
Truy xuất một phần dữ liệu từ Tensor nhiều chiều.

**Yêu cầu:**
- Cho biết shape của `a[1]` và `a[1:, 2:4]`.
- Thử lấy hai phần tử cuối cùng trên chiều thứ ba.


In [6]:
import torch
torch.manual_seed(0)

a = torch.randn(3, 4, 5)
print("Original tensor:")
print(a)
print("shape cua a:", a.shape)

print("\nTensor a[1]:")
print(a[1])
print("shape cua a[1]:", a[1].shape)

print("\nTensor a[1:, 2:4]:")
print(a[1:, 2:4])
print("shape cua a[1:, 2:4]:", a[1:, 2:4].shape)

# --- Lay hai phan tu cuoi cung tren chieu thu ba (dim=2) ---
last_two = a[:, :, -2:]
print("\nHai phan tu cuoi tren chieu thu ba a[:, :, -2:]:")
print(last_two)
print("shape:", last_two.shape)


Original tensor:
tensor([[[-1.1258, -1.1524, -0.2506, -0.4339,  0.8487],
         [ 0.6920, -0.3160, -2.1152,  0.3223, -1.2633],
         [ 0.3500,  0.3081,  0.1198,  1.2377,  1.1168],
         [-0.2473, -1.3527, -1.6959,  0.5667,  0.7935]],

        [[ 0.5988, -1.5551, -0.3414,  1.8530,  0.7502],
         [-0.5855, -0.1734,  0.1835,  1.3894,  1.5863],
         [ 0.9463, -0.8437, -0.6136,  0.0316, -0.4927],
         [ 0.2484,  0.4397,  0.1124,  0.6408,  0.4412]],

        [[-0.1023,  0.7924, -0.2897,  0.0525,  0.5943],
         [ 1.5419,  0.5073, -0.5910, -0.5692,  0.9200],
         [ 1.1108,  1.2899, -1.4959, -0.1938,  0.4455],
         [ 1.3253, -1.6293, -0.5497, -0.4798, -0.4997]]])
shape cua a: torch.Size([3, 4, 5])

Tensor a[1]:
tensor([[ 0.5988, -1.5551, -0.3414,  1.8530,  0.7502],
        [-0.5855, -0.1734,  0.1835,  1.3894,  1.5863],
        [ 0.9463, -0.8437, -0.6136,  0.0316, -0.4927],
        [ 0.2484,  0.4397,  0.1124,  0.6408,  0.4412]])
shape cua a[1]: torch.Size([4, 5])


**Nhận xét Bài 4:**
- `a` có shape (3, 4, 5).
- `a[1]` lấy lát cắt tại index 1 của chiều 0 -> shape (4, 5) (mất 1 chiều).
- `a[1:, 2:4]` lấy chiều 0 từ 1 đến hết (2 phần tử), chiều 1 lấy index 2,3 (2 phần tử), giữ nguyên chiều 2 (5 phần tử) -> shape (2, 2, 5).
- Để lấy 2 phần tử cuối trên chiều thứ ba: `a[:, :, -2:]` -> shape (3, 4, 2).


---
## Bài 5. Reshape và Flatten
Thay đổi hình dạng Tensor mà không làm thay đổi số phần tử.

**Yêu cầu:**
- Kiểm tra số phần tử trước và sau khi reshape.
- Thử reshape Tensor thành kích thước 2 x 6.


In [7]:
import torch
torch.manual_seed(0)

a = torch.randn(3, 4)
print("Original:")
print(a)
print("shape:", a.shape, "| numel:", a.numel())

print("\nFlatten:")
print(a.ravel())
print("shape flatten:", a.ravel().shape)

print("\nReshape to 3 x 2 x 2:")
b = a.reshape(3, 2, 2)
print(b)
print("shape:", b.shape, "| numel:", b.numel())

# --- Yeu cau them ---
print("\nReshape to 2 x 6:")
c = a.reshape(2, 6)
print(c)
print("shape:", c.shape, "| numel:", c.numel())

assert a.numel() == b.numel() == c.numel() == 12, "reshape phai giu nguyen so phan tu!"


Original:
tensor([[ 1.5410, -0.2934, -2.1788,  0.5684],
        [-1.0845, -1.3986,  0.4033,  0.8380],
        [-0.7193, -0.4033, -0.5966,  0.1820]])
shape: torch.Size([3, 4]) | numel: 12

Flatten:
tensor([ 1.5410, -0.2934, -2.1788,  0.5684, -1.0845, -1.3986,  0.4033,  0.8380,
        -0.7193, -0.4033, -0.5966,  0.1820])
shape flatten: torch.Size([12])

Reshape to 3 x 2 x 2:
tensor([[[ 1.5410, -0.2934],
         [-2.1788,  0.5684]],

        [[-1.0845, -1.3986],
         [ 0.4033,  0.8380]],

        [[-0.7193, -0.4033],
         [-0.5966,  0.1820]]])
shape: torch.Size([3, 2, 2]) | numel: 12

Reshape to 2 x 6:
tensor([[ 1.5410, -0.2934, -2.1788,  0.5684, -1.0845, -1.3986],
        [ 0.4033,  0.8380, -0.7193, -0.4033, -0.5966,  0.1820]])
shape: torch.Size([2, 6]) | numel: 12


**Nhận xét Bài 5:**
- Tensor gốc 3 x 4 = 12 phần tử.
- `ravel()` / `flatten()` trải phẳng thành vector 12 phần tử.
- `reshape(3, 2, 2)` và `reshape(2, 6)` đều có 12 phần tử — reshape chỉ đổi hình dạng, không đổi dữ liệu/tổng số phần tử.
- Điều kiện reshape hợp lệ: tích các kích thước mới phải bằng `numel` gốc.


---
## Bài 6. Phép toán và nhân ma trận
Thực hiện các phép toán phần tử và phép nhân ma trận bằng toán tử `@`.

**Yêu cầu:**
- Phân biệt phép toán `a * b` với `a @ b.T`.
- Cho biết shape của kết quả `a @ b.T`.


In [8]:
import torch
torch.manual_seed(0)

a = torch.randn(2, 3)
b = torch.randn(2, 3)
print("a:")
print(a)
print("b:")
print(b)

print("\na + b:")
print(a + b)

print("\na / b:")
print(a / b)

print("\na^2:")
print(a ** 2)

print("\na * b (nhan tung phan tu):")
print(a * b)
print("shape:", (a * b).shape)

print("\nMatrix multiplication a @ b.T:")
print(a @ b.T)
print("shape:", (a @ b.T).shape)


a:
tensor([[ 1.5410, -0.2934, -2.1788],
        [ 0.5684, -1.0845, -1.3986]])
b:
tensor([[ 0.4033,  0.8380, -0.7193],
        [-0.4033, -0.5966,  0.1820]])

a + b:
tensor([[ 1.9443,  0.5446, -2.8980],
        [ 0.1651, -1.6812, -1.2166]])

a / b:
tensor([[ 3.8205, -0.3501,  3.0292],
        [-1.4093,  1.8177, -7.6831]])

a^2:
tensor([[2.3747, 0.0861, 4.7471],
        [0.3231, 1.1762, 1.9561]])

a * b (nhan tung phan tu):
tensor([[ 0.6216, -0.2459,  1.5671],
        [-0.2293,  0.6471, -0.2546]])
shape: torch.Size([2, 3])

Matrix multiplication a @ b.T:
tensor([[ 1.9428, -0.8431],
        [ 0.3264,  0.1632]])
shape: torch.Size([2, 2])


**Nhận xét Bài 6:**
- `a * b` là **nhân từng phần tử (element-wise)**: yêu cầu cùng shape (hoặc broadcast được). Với `a, b: (2,3)` -> kết quả `(2, 3)`.
- `a @ b.T` là **nhân ma trận**: `a:(2,3)` nhân `b.T:(3,2)` -> kết quả `(2, 2)`. Mỗi phần tử là tích vô hướng của một hàng `a` với một hàng `b`.
- Công thức shape: (m x n) @ (n x p) -> (m x p). Ở đây (2x3)@(3x2)->(2x2).


---
## Bài 7. Các hàm thống kê trên Tensor
Tính trung bình, độ lệch chuẩn và tổng tích lũy theo từng cột.

**Yêu cầu:** Thay `dim=0` bằng `dim=1` và nhận xét sự khác nhau.


In [9]:
import torch
torch.manual_seed(0)

a = torch.randn(3, 4)
print(a)

print("\nMean (dim=0, theo cot):")
print(torch.mean(a, dim=0))

print("\nStandard deviation (dim=0):")
print(torch.std(a, dim=0))

print("\nCumulative sum (dim=0):")
print(torch.cumsum(a, dim=0))

print("\n===== So sanh voi dim=1 (theo hang) =====")
print("\nMean (dim=1):")
print(torch.mean(a, dim=1))

print("\nStandard deviation (dim=1):")
print(torch.std(a, dim=1))

print("\nCumulative sum (dim=1):")
print(torch.cumsum(a, dim=1))


tensor([[ 1.5410, -0.2934, -2.1788,  0.5684],
        [-1.0845, -1.3986,  0.4033,  0.8380],
        [-0.7193, -0.4033, -0.5966,  0.1820]])

Mean (dim=0, theo cot):
tensor([-0.0876, -0.6985, -0.7907,  0.5295])

Standard deviation (dim=0):
tensor([1.4222, 0.6088, 1.3020, 0.3297])

Cumulative sum (dim=0):
tensor([[ 1.5410, -0.2934, -2.1788,  0.5684],
        [ 0.4565, -1.6920, -1.7754,  1.4065],
        [-0.2628, -2.0954, -2.3721,  1.5885]])

===== So sanh voi dim=1 (theo hang) =====

Mean (dim=1):
tensor([-0.0907, -0.3104, -0.3843])

Standard deviation (dim=1):
tensor([1.5809, 1.0972, 0.3993])

Cumulative sum (dim=1):
tensor([[ 1.5410,  1.2476, -0.9312, -0.3628],
        [-1.0845, -2.4831, -2.0798, -1.2417],
        [-0.7193, -1.1226, -1.7192, -1.5372]])


**Nhận xét Bài 7:**
- Với Tensor shape (3, 4) (3 hàng, 4 cột):
  - `dim=0`: tính dọc theo các **hàng** -> gom theo **cột**. Kết quả shape (4,) (1 giá trị cho mỗi cột). `cumsum` cộng dồn từ trên xuống dưới.
  - `dim=1`: tính dọc theo các **cột** -> gom theo **hàng**. Kết quả shape (3,) (1 giá trị cho mỗi hàng). `cumsum` cộng dồn từ trái sang phải.
- Quy tắc nhớ: `dim=k` là chiều bị triệt tiêu (collapse) sau khi tính mean/std.


---
## Bài 8. Autograd cơ bản
Tạo biến có `requires_grad=True` và dùng `backward()` để tính đạo hàm tự động.

**Yêu cầu:**
- Tính thủ công đạo hàm của y = x^2 tại x = 3.6 và so sánh với `x.grad`.
- Đổi hàm thành y = 3*x**2 + 2*x + 1 và kiểm tra gradient.


In [10]:
import torch

x = torch.tensor(
    3.6,
    requires_grad=True
)
y = x * x
y.backward()
print("x =", x)
print("y =", y)
print("x.grad =", x.grad)

# --- Tinh thu cong: y = x^2 -> dy/dx = 2x = 2*3.6 = 7.2 ---
print("Tinh thu cong dy/dx = 2*3.6 =", 2*3.6)

# --- Yeu cau them: y = 3*x**2 + 2*x + 1 ---
x2 = torch.tensor(3.6, requires_grad=True)
y2 = 3*x2**2 + 2*x2 + 1
y2.backward()
print("\nVoi y = 3*x**2 + 2*x + 1 tai x=3.6:")
print("y2 =", y2.item())
print("x2.grad (autograd) =", x2.grad.item())
print("Tinh thu cong dy/dx = 6x+2 =", 6*3.6+2)


x = tensor(3.6000, requires_grad=True)
y = tensor(12.9600, grad_fn=<MulBackward0>)
x.grad = tensor(7.2000)
Tinh thu cong dy/dx = 2*3.6 = 7.2

Voi y = 3*x**2 + 2*x + 1 tai x=3.6:
y2 = 47.07999801635742
x2.grad (autograd) = 23.599998474121094
Tinh thu cong dy/dx = 6x+2 = 23.6


**Nhận xét Bài 8:**
- y = x^2 => dy/dx = 2x. Tại x=3.6: 2*3.6=7.2 khớp với `x.grad`.
- y = 3x^2+2x+1 => dy/dx = 6x+2. Tại x=3.6: 6*3.6+2=23.6 khớp với autograd.
- `requires_grad=True` yêu cầu PyTorch lưu đồ thị tính toán; `backward()` lan truyền ngược để tính gradient.


---
## Bài 9. Gradient Descent tìm hệ số đa thức
Sinh dữ liệu từ đa thức y = x^2 + 2x + 3, sau đó dùng NAdam và autograd để tìm lại ba hệ số.

**Yêu cầu:**
- So sánh ba hệ số học được với `[1, 2, 3]`.
- Giải thích vai trò của `optimizer.zero_grad()`, `mse.backward()` và `optimizer.step()`.


In [11]:
import numpy as np
import torch
np.random.seed(0)
torch.manual_seed(0)

polynomial = np.poly1d([1, 2, 3])
N = 20
X = np.random.randn(N, 1) * 5
Y = polynomial(X)

XX = np.hstack([
    X * X,
    X,
    np.ones_like(X)
])

w = torch.randn(3, 1, requires_grad=True)
x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y, dtype=torch.float32)

optimizer = torch.optim.NAdam([w], lr=0.01)
print("Initial coefficients:")
print(w)

for epoch in range(1000):
    optimizer.zero_grad()
    y_pred = x @ w
    mse = torch.mean(torch.square(y - y_pred))
    mse.backward()
    optimizer.step()

print("Learned coefficients:")
print(w)
print("He so thuc: [1, 2, 3]")


Initial coefficients:
tensor([[ 1.5410],
        [-0.2934],
        [-2.1788]], requires_grad=True)
Learned coefficients:
tensor([[1.0017],
        [2.0058],
        [2.8337]], requires_grad=True)
He so thuc: [1, 2, 3]


**Nhận xét Bài 9:**
- Ba hệ số học được xấp xỉ [1, 2, 3] (sai số nhỏ do làm tròn / số epoch hữu hạn). Mô hình hồi quy tuyến tính trên đặc trưng [x^2, x, 1] đã khôi phục đúng đa thức sinh dữ liệu.
- Vai trò 3 lệnh trong vòng lặp:
  - `optimizer.zero_grad()`: xóa gradient cũ (PyTorch cộng dồn gradient, nên phải reset mỗi epoch).
  - `mse.backward()`: tính gradient của loss theo `w` bằng autograd (lan truyền ngược).
  - `optimizer.step()`: cập nhật `w` theo hướng ngược gradient (NAdam là biến thể Gradient Descent có momentum + lr thích nghi).


---
## Bài 10. Giải hệ phương trình bằng Autograd
Tối thiểu hóa tổng sai số bình phương để tìm A, B, C, D thỏa bốn phương trình.

**Yêu cầu:**
- Kiểm tra nghiệm thu được với bốn phương trình: A+B=9, C-D=1, A+C=8, B-D=2.
- Nhận xét vì sao bài toán có thể có nhiều nghiệm.


In [12]:
import random
import torch
random.seed(0)
torch.manual_seed(0)

A = torch.tensor(random.random(), requires_grad=True)
B = torch.tensor(random.random(), requires_grad=True)
C = torch.tensor(random.random(), requires_grad=True)
D = torch.tensor(random.random(), requires_grad=True)

EPOCHS = 2000
optimizer = torch.optim.NAdam([A, B, C, D], lr=0.01)

for _ in range(EPOCHS):
    y1 = A + B - 9
    y2 = C - D - 1
    y3 = A + C - 8
    y4 = B - D - 2
    sqerr = y1*y1 + y2*y2 + y3*y3 + y4*y4
    optimizer.zero_grad()
    sqerr.backward()
    optimizer.step()

print("A =", A)
print("B =", B)
print("C =", C)
print("D =", D)

# --- Kiem tra nghiem ---
print("\nKiem tra:")
print(f"A+B = {(A+B).item():.4f} (mong doi 9)")
print(f"C-D = {(C-D).item():.4f} (mong doi 1)")
print(f"A+C = {(A+C).item():.4f} (mong doi 8)")
print(f"B-D = {(B-D).item():.4f} (mong doi 2)")


A = tensor(4.5494, requires_grad=True)
B = tensor(4.4506, requires_grad=True)
C = tensor(3.4506, requires_grad=True)
D = tensor(2.4506, requires_grad=True)

Kiem tra:
A+B = 9.0000 (mong doi 9)
C-D = 1.0000 (mong doi 1)
A+C = 8.0000 (mong doi 8)
B-D = 2.0000 (mong doi 2)


**Nhận xét Bài 10:**
- Nghiệm thỏa mãn cả 4 phương trình với sai số ~0 (ví dụ A~3, B~6, C~5, D~4 tùy lần chạy — học viên đối chiếu số in ra).
- Bài toán **có thể có nhiều nghiệm** vì 4 phương trình không độc lập tuyến tính: (A+B)-(A+C)+(C-D)-(B-D) = 0恒成立 với mọi A,B,C,D, nên hạng của hệ < 4. Hệ phụ thuộc tuyến tính nên có 1 bậc tự do -> vô số nghiệm. Gradient descent sẽ hội tụ về một nghiệm gần điểm khởi tạo ngẫu nhiên.


---
## 3. Bài tập áp dụng
### Bài 11. Viết chương trình PyTorch để thực hiện các yêu cầu sau:
- Tạo Tensor X có 24 phần tử và reshape lần lượt thành 4 x 6, 2 x 12 và 2 x 3 x 4.
- Tạo hai Tensor A, B kích thước 3 x 4; tính A+B, A*B, A/B, trung bình theo cột và độ lệch chuẩn theo cột.
- Cho y = 2x^2 + 5x + 3. Dùng autograd tính dy/dx tại x = 2 và kiểm tra bằng phép tính thủ công.
- Thay đa thức trong Bài 9 bằng y = 2x^2 - 3x + 5; huấn luyện và cho biết các hệ số tìm được.


#### Bài 11.1. Reshape Tensor 24 phần tử


In [13]:
import torch

X = torch.arange(24)  # 24 phan tu 0..23
print("X:", X)
print("numel:", X.numel())

X_4x6 = X.reshape(4, 6)
X_2x12 = X.reshape(2, 12)
X_2x3x4 = X.reshape(2, 3, 4)

print("\n4x6:")
print(X_4x6)
print("\n2x12:")
print(X_2x12)
print("\n2x3x4:")
print(X_2x3x4)


X: tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 23])
numel: 24

4x6:
tensor([[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17],
        [18, 19, 20, 21, 22, 23]])

2x12:
tensor([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]])

2x3x4:
tensor([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]])


#### Bài 11.2. Phép toán và thống kê theo cột


In [14]:
import torch
torch.manual_seed(0)

A = torch.randn(3, 4)
B = torch.randn(3, 4)
print("A:")
print(A)
print("B:")
print(B)

print("\nA+B:")
print(A + B)
print("\nA*B (element-wise):")
print(A * B)
print("\nA/B:")
print(A / B)

print("\nTrung binh theo cot (dim=0):")
print(torch.mean(A, dim=0))
print("\nDo lech chuan theo cot (dim=0):")
print(torch.std(A, dim=0))


A:
tensor([[ 1.5410, -0.2934, -2.1788,  0.5684],
        [-1.0845, -1.3986,  0.4033,  0.8380],
        [-0.7193, -0.4033, -0.5966,  0.1820]])
B:
tensor([[-0.8567,  1.1006, -1.0712,  0.1227],
        [-0.5663,  0.3731, -0.8920, -1.5091],
        [ 0.3704,  1.4565,  0.9398,  0.7748]])

A+B:
tensor([[ 0.6843,  0.8072, -3.2500,  0.6911],
        [-1.6508, -1.0255, -0.4886, -0.6711],
        [-0.3489,  1.0532,  0.3432,  0.9569]])

A*B (element-wise):
tensor([[-1.3201, -0.3229,  2.3339,  0.0697],
        [ 0.6142, -0.5218, -0.3598, -1.2647],
        [-0.2664, -0.5875, -0.5607,  0.1411]])

A/B:
tensor([[-1.7988, -0.2666,  2.0340,  4.6326],
        [ 1.9150, -3.7484, -0.4522, -0.5553],
        [-1.9419, -0.2769, -0.6348,  0.2349]])

Trung binh theo cot (dim=0):
tensor([-0.0876, -0.6985, -0.7907,  0.5295])

Do lech chuan theo cot (dim=0):
tensor([1.4222, 0.6088, 1.3020, 0.3297])


#### Bài 11.3. Autograd với y = 2x^2 + 5x + 3 tại x=2


In [15]:
import torch

x = torch.tensor(2.0, requires_grad=True)
y = 2*x**2 + 5*x + 3
y.backward()
print("y =", y.item())
print("dy/dx (autograd) =", x.grad.item())
print("Kiem tra thu cong: dy/dx = 4x+5 =", 4*2+5)


y = 21.0
dy/dx (autograd) = 13.0
Kiem tra thu cong: dy/dx = 4x+5 = 13


#### Bài 11.4. Gradient Descent với y = 2x^2 - 3x + 5


In [16]:
import numpy as np
import torch
np.random.seed(1)
torch.manual_seed(1)

polynomial = np.poly1d([2, -3, 5])  # y = 2x^2 - 3x + 5
N = 20
X = np.random.randn(N, 1) * 5
Y = polynomial(X)

XX = np.hstack([X*X, X, np.ones_like(X)])
w = torch.randn(3, 1, requires_grad=True)
x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y, dtype=torch.float32)

optimizer = torch.optim.NAdam([w], lr=0.01)
print("Initial coefficients:")
print(w)

for epoch in range(1000):
    optimizer.zero_grad()
    y_pred = x @ w
    mse = torch.mean(torch.square(y - y_pred))
    mse.backward()
    optimizer.step()

print("Learned coefficients (mong doi ~[2, -3, 5]):")
print(w)


Initial coefficients:
tensor([[0.6614],
        [0.2669],
        [0.0617]], requires_grad=True)
Learned coefficients (mong doi ~[2, -3, 5]):
tensor([[ 2.0372],
        [-2.9304],
        [ 2.4373]], requires_grad=True)


---
**--- Hết ---**
